### PR TITLE
Implement kernel for casting float to decimal

### PR DESCRIPTION
**Pull request by ttnghia**
This implements a casting operation from floating point values to decimal values, following the rules of Apache Spark. The casting function also returns an additional boolean flag indicating if there was any failures with any input rows, leaving the caller to make decision whether to accept the result as-is or raise an exception.

### CODE REVIEW

**Comment by ttnghia** [comment]
Related: https://github.com/NVIDIA/spark-rapids/issues/9847

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

File: src/main/java/com/nvidia/spark/rapids/jni/DecimalUtils.java [hunk 1]
```diff
@@ -177,6 +180,38 @@ public static Table add128(ColumnView a, ColumnView b, int targetScale) {
     return new Table(add128(a.getNativeView(), b.getNativeView(), targetScale));
   }
 
+  /**
+   * A class to store the result of a cast operation from floating point values to decimals.
+   * <p>
+   * Since the result column may or may not be used regardless of the value of hasFailure, we
+   * need to kept it and let the caller to decide.
```
Comment (jlowe) [inline src/main/java/com/nvidia/spark/rapids/jni/DecimalUtils.java:187]:
```suggestion
   * need to keep it and let the caller to decide.
```

File: src/test/java/com/nvidia/spark/rapids/jni/DecimalUtilsTest.java [hunk 2]
```diff
@@ -192,9 +193,9 @@ void simplePosDivOneByZero() {
     try (ColumnVector lhs =
              makeDec128Column("1.0", "10.0", "1.0", "1000000000000000000000000000000000000.0");
          ColumnVector rhs =
-             makeDec128Column("1",   "2",    "0",   "5");
+             makeDec128Column("1", "2", "0", "5");
          ColumnVector expectedBasic =
-             makeDec128Column("1.0", "5.0",  "0",   "200000000000000000000000000000000000.0");
+             makeDec128Column("1.0", "5.0", "0", "200000000000000000000000000000000000.0");
```
Comment (jlowe) [inline src/test/java/com/nvidia/spark/rapids/jni/DecimalUtilsTest.java:198]: Nit: The whitespace being removed here was for readability to line up the row data across columns, and this unrelated change makes it arguably harder to read.

**Review by jlowe — APPROVED** [review]
Minor nits on comments.  Do we have any performance measurements for the new kernel?  Wondering if it's worth making a benchmark for it.

**Comment by ttnghia** [comment]
> Minor nits on comments. Do we have any performance measurements for the new kernel? Wondering if it's worth making a benchmark for it.

Previously @thirtiseven helped with benchmarking, like this https://github.com/NVIDIA/spark-rapids/pull/10917#issuecomment-2134569776. @thirtiseven can you run such benchmark again on this please? Thanks.

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Comment by thirtiseven** [comment]
New perf numbers:

| Type | CPU | 24.08 | gluten |  This PR |
| --- | --- | --- | --- | --- |
| Double | 147,121.5 | 524.25 | 7,105.5 | 520.5 |
| Float | 89,442 | 440.75 | 8,007.5 | 657.75 |

Float type is a little slower than current code, but it runs much faster than cpu and gluten.
